---
title: "Analysis of endometrial cancer"
author: "Sergio Vela Moreno"
date: "2023-01-07"
output: html_document
---

# Analysis of endometrial cancer

Firstly, the work directory and a seed to ensure reproducibility are set:

```{r}
setwd("C:/Users/Pc/Desktop/PhDPosition")
set.seed(1234)
```

Then the data is loaded into the R environment:

```{r}
##Example to load the dataset.
##UCEC miRNA-Seq data: miRNA isoform expression and phenodata.
a <- load("TCGA-UCEC_miRNA_isoform_expression.RData")
a #"ucec"
##check dimensions of the raw count
dim(ucec$count) #2243  571

dim(ucec$pheno) #571   2

##this dataset can be used for further differential expression analysis,
#visualization using t-SNE, etc.
```
## Determine the number of different cancer stages within the samples:

In order to determine the number of different cancer stages within the samples, the phenotypes in the *ucec* object must be explored:

```{r}
stages <- levels(ucec$pheno$group)
```

```{r}
library(ggplot2)

ggplot()+
  geom_bar(data=ucec$pheno, aes(group, fill=factor(group)))+
  ggtitle("Samples distribution in cancer stages")
```

In total there are 4 different cancer stages and some normal samples: according to the factor levels there are 6 different stages. However, when the distribution of patients is represented in the different groups, it is shown that there are no patients belonging to the group `stage_I`.

## Perform cluster analysis using t-sne and visualize how the miRNA samples are clustered according to cancer stages:

Data cleaning must be done in order to eliminate `NA` or missing values before the *t-sne* is performed. It is important to use the transpose matrix in this case, as the different samples are represented in the columns of the original counts data matrix and t-sne applies dimension reduction to the variables in columns.

```{r}
ucec_matrix <- as.matrix(ucec$count)
ucec_matrix <- na.omit(ucec_matrix)

Matrix_t <- t(ucec_matrix)
```

```{r}
library(Rtsne)

Rtsne_output <- Rtsne(Matrix_t,pca=FALSE,perplexity=30,theta=0.0, max_iter = 3000)
```

After t-sne has been performed, its resulting output is represented in order to visualize how miRNA samples are clustered:

```{r}
Y <- as.data.frame(Rtsne_output$Y)

ggplot(data = Y, aes(x=V1, y=V2, col=ucec$pheno$group))+
  geom_point()+
  labs(x="t-sne 1", y="t-sne 2", color="Cancer stages")+
  ggtitle("t-sne representation")
```

According to the t-sne plot, it might be possible to differentiate cancerous and normal samples using miRNA, but samples in different cancer stages are harder to cluster.

## Perform pairwise differential miRNA expression analysis:

The next step is to perform a pairwise differential miRNA expression analysis. For this purpose, the package `DESeq2` will be used. This package needs a `DESeqDataSet` (dds), which in this case is obtained from the counts matrix and the group variable from the original data.

```{r}
factores <- ucec$pheno
rownames(factores) <- factores$samplenames

dds <- DESeqDataSetFromMatrix(countData = ucec_matrix,
                              colData = factores,
                              design = ~ group)
```

Once the dds object has been obtained, differential expression analysis is performed.

```{r}
DEAResults <-  DESeq(dds)
```

As a pairwise differential miRNA expression analysis is required, when displaying results it is needed to indicate the conditions that will be compared in each case:

```{r}
res1 <- results(DEAResults, contrast = c("group", "Normal", "StageI")) #Normal-StageI
res2 <- results(DEAResults, contrast = c("group", "Normal", "StageII")) #Normal-StageII
res3 <- results(DEAResults, contrast = c("group", "Normal", "stage_III")) #Normal-StageIII
res4 <- results(DEAResults, contrast = c("group", "Normal", "StageIV")) #Normal-StageIV
res5 <- results(DEAResults, contrast = c("group", "StageI", "StageII")) #StageI-StageII
res6 <- results(DEAResults, contrast = c("group", "StageI", "stage_III")) #StageI-StageIII
res8 <- results(DEAResults, contrast = c("group", "StageII","stage_III")) #StageII-StageIII
res9 <- results(DEAResults, contrast = c("group", "StageII", "StageIV")) #StageII-StageIV
res10 <- results(DEAResults, contrast = c("group","stage_III","StageIV")) #StageIII-StageIV
```

Once pairwise differential miRNA expression analysis is performed, the resulting DEGs can be identified based on their p-adjusted value. In this case the cut off to consider a gene as differentially expressed is a p-value lower than 0.05.
The following chunk represents the example for the identification of DEGs between normal and StageIV cancerous samples. These have been the two conditions chosen for comparison because they have a lower number of samples and therefore its following visualization will be clearer, but the same process can be followed with the other cancer stages as well.

```{r}
res4$threshold <- as.logical(res4$padj < 0.05)

RelevantDEGsExpression <- ucec_matrix[res4$threshold,]

RelevantDEGs <- rownames(RelevantDEGsExpression)
```

As a result, 163 miRNA codifying genes have been identified as differentially expressed between both conditions.

## Visualize the differentially expressed miRNAs:

The expression of these DEGs can be represented using different visualization methods, such as heatmaps. As many DEGs have been identified, only a few of them are represented in order to facilitate heatmap interpretation.
On top of that, data filtering has been performed in order to only represent gene expression of samples corresponding to the two conditions compared (Normal ans StageIV samples).

```{r}
library(dplyr)

NormalSamples <- rownames(factores %>%  #Identify normal samples IDs
  dplyr::filter( group == 'Normal') %>%
  dplyr::select(samplenames))

StageIVSamples <- rownames(factores %>%  #Identify StageIV samples IDs
                   dplyr::filter( group == 'StageIV') %>%
                   dplyr::select(samplenames))

RelevantDEGsExpressionNormal <- RelevantDEGsExpression[1:10,NormalSamples]
RelevantDEGsExpressionStageIV <- RelevantDEGsExpression[1:10,StageIVSamples]

RelevantDEGsExpressionSample <- cbind(RelevantDEGsExpressionNormal,RelevantDEGsExpressionStageIV)
```

After gene expression of normal and StageIV samples has been integrated in a simplified dataframe, it is represented using a heatmap.

```{r}
library("pheatmap")

pheatmap(RelevantDEGsExpressionSample,
         color = ,
         fontsize = 6,
         show_rownames = T,
         show_colnames = T,
         cluster_cols = T,
         cluster_rows = F,
         main = "Heatmap Control-StageIV")
```

According to the resulting figure, there are significant differences in the expression of considered genes (represented in the rows of the heatmap) between normal and cancerous samples (represented in the columns). In particular, within the first 5 genes represented (first five rows) there is an upregulation of gene expression in cancerous samples when compared to normal samples.

## Identify target genes for differentially expressed miRNAs using any of the target identification methods and perform KEGG pathway and GO enrichment analysis:

Afterwards, the miRNAs identified as DEGs are used to identify their target genes using the package `multiMiR`. However, due to computational limitations, target genes were identified just for one of the miRNAs.

```{r}
library(multiMiR)

TargetGenes <- get_multimir(mirna = 'hsa-let-7a-5p', summary = TRUE)

#TargetGenesTotal <- get_multimir(mirna = RelevantDEGs, summary = TRUE) #This code is used to identify target genes for all miRNAs, but is highly time-consuming.
```

As a result, 3008 genes have been identified as target genes for the miRNA *hsa-let-7a-5p*.
These genes can be used to perform a Gene Ontology and KEGG Enrichment Analysis.
First, a GO enrichment anylisis is performed to identify GO terms related to the target genes.

```{r}
library(org.Hs.eg.db)
library(AnnotationDbi)
library(clusterProfiler)

GenesList <- TargetGenes@summary$target_ensembl

GO_results <- enrichGO(gene=GenesList,OrgDb="org.Hs.eg.db",keyType="ENSEMBL",ont="BP")

MostRelatedGOTerms <- plot(barplot(GO_results,showCategory = 15, font.size = 7.5))
```

After GO enrichment, the GO terms which are more related to the identified target genes are represented. Among the targets of miRNAs, genes related to protein ubiquitinization, cellular response to external stimulus and regulation of autophagy can be found. From a biological point of view the downexpression of these genes due to miRNAs binding can lead to lower degradation of cell proteins, an uncontrolled proliferation of cells because of the reduced response to external stimuli and reduced cell death.

Finally, KEGG Enrichment Analysis is performed, identifying the KEGG pathways which are related to the target genes.

```{r}
GenesTranslate <- bitr(GenesList, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = 'org.Hs.eg.db')

KEGG_genes <- enrichKEGG(gene = GenesTranslate$ENTREZID, organism = "hsa", pvalueCutoff = 0.05)

TopRelatedPathways <- plot(barplot(KEGG_genes,showCategory = 15, font.size = 7.5))
```

As a result, among the most related pathways there are pathways involved with cell proliferation such as Cell cycle, Proteoglycans in cancer and Viral carcinogenesis. 
It is also remarkable the relation of target genes with pathways involved in neurodegenerative diseases proliferation and different viruses infection.

```{r}
library(enrichplot)
library(ggnewscale)

KEGG_genesPr <- pairwise_termsim(KEGG_genes)
enrichplot::emapplot(KEGG_genesPr, showCategory = 10)
```

Identified relevant pathways can also be plotted with an enrichment plot to identify if there are any relevant relationships within them. As a result, interactions between cell cycle, viral carcinogenesis, different viruses infection and Hepatitis B (a virus dependent disease) have been proven.